### PR TITLE
30 implement a bayesian opimization algorithm for cadet process based on axbotorch

### PR DESCRIPTION
**Pull request by flo-schu**
implements the developer API of Ax (https://ax.dev/tutorials/gpei_hartmann_developer.html) for CADET-Process. 

## new features

- [x] single objective optimization
- [x] multi-objective optimization
- [x] implement different algorithms for multi-objective optimization https://ax.dev/tutorials/multiobjective_optimization.html
  - [x] Subclass AxInterface for different algorithms
  - [x] GPEI
- ~~termination criteria (nice to have)~~
  - ~~multi-objective (can be singular in the case of minimization)~~
- [x] linear constraints
- [x] nonlinear constraints
- [x] unittests 
  - [x] single objective NRMSE
  - [x] multi-objective (fit gaussian curve)
- [ ] parallelize evaluations https://ax.dev/tutorials/scheduler.html
- [ ] docstrings
- [x] Parametrisierung (sollte einstellbar sein - API - genaue Werte über Forum lösen)
  - [x] SEM: Forumsbeitrag um Wissen zu sammeln (sollte einstellbar sein)
  - [x] ~~hyperparameters: Forumsbeitrag exploration vs exploitation was sind gute Parameter für verschiedene Algortihmen~~

## improvements

- [x] increases the flexibility of setup_optimization_problem by allowing to pass custom obj_functions

## infra
- [x] Make ax optional dependency

### CODE REVIEW

**Comment by flo-schu** [comment]
@schmoelder: Right now multiobjective optimization works. Today I would work on the following items

A question for you: Do you think you can work on making intermediate results of the evaluation-toolchain available, or give me a short Idea on how to do this? This would improve the efficiency of my callbacks, unless I'm missing something again :smile:

## to-dos

1. finalize unittests for
  - [ ] multi-objective
  - [ ] single-objective --> this involves making single-objective work again (since it was broken for multi-objective)

because these follow different implementation. I will see if I can implement single obj. as a special case of multi-obj. or if I have to use different implementations dependent on the case

2. Ax provides a runner class in the developer API

- [ ] I will try if simulation results can be performed in the runner and saved as meta information and accessed in the metrics? This way, computation would only run once. Less overhead.

3. Start writing docstrings for the classes and methods and clean up code

4. If I get to it, switch from SOBOL to HOPSY for initial evaluations

**Comment by flo-schu** [comment]
## Tasks for this week

**priority**
- [ ] continue testing sweet for Ax. As benchmarking problems that are reproducibly solved
  - [ ] set seed in for MCMC sampling in Ax according to https://github.com/facebook/Ax/issues/1663#issuecomment-1594881392
  - [ ] also use for scipy
  - [ ] pymoo (this week optional)

- [ ] work on models in Ax
  - [ ] implement different acquisition functions / surrogate models --> model bridge- [ ] fix linear constraints problem in scipy
  - [ ] hopsy for initial sampling

**Parallel tasks**
- [ ] fit GP (pot. Scipy) on existing population of parameter samples to generate visualizations
  - [ ] partial dependence plots
  - [ ] ...

**Comment by schmoelder** [comment]
If I can add:
- Allow custom initialization (e.g. using sample from hopsy)

**Comment by schmoelder** [comment]
I just use [np.testing.almost_equal](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_almost_equal.html).

**Comment by flo-schu** [comment]
@schmoelder do you know a good simple multi-objective optimization problem with linear constraints and a known pareto front, ideally implemented in python (not ax so I can use it for other optimization packages as well)? 

If not, have you experience how to derive the pareto front for simple problems or do you know who to ask.

I asked chatGPT for a problem but I suspect it got the pareto front wrong. But maybe its also me and we can have a look tomorrow together

**Comment by schmoelder** [comment]
Maybe some of pymoo's test problems would work (see: https://pymoo.org/problems/test_problems.html#Multi-Objective).

**Comment by flo-schu** [comment]
@schmoelder i have written a test suite for Ax. This is not completely finalized, but I think its a good start to use this also for other optimizers. I wouldn't mind if you took a look at it and review if you're happy with the implementation. I don't mind changing parts of the implementation.

**Comment by schmoelder** [comment]
Thanks a lot! I will definitely have a look and review!

Btw, I took the liberty to rebase again. you might have to `reset --hard` (but, as always, be careful with that one! ;-) )

**Comment by flo-schu** [comment]
> you might have to `reset --hard` (but, as always, be careful with that one! ;-) )

I used `git merge --abort` when the pull was not successful and updated strategy for reconciling diverging branches. This worked quite smooth

**Comment by schmoelder** [comment]
This comment might get us some idea on how to address custom initialization:

https://github.com/facebook/Ax/issues/768#issuecomment-1005169890

File: CADETProcess/optimization/axAdapater.py [hunk 1]
```diff
@@ -0,0 +1,498 @@
...
+from botorch.acquisition.analytic import UpperConfidenceBound, NoisyExpectedImprovement
+from ax.models.torch.botorch_modular.model import BoTorchModel
+from ax.modelbridge import TorchModelBridge, ModelBridge
+from ax.modelbridge.registry import Cont_X_trans, Y_trans
+
+from CADETProcess.dataStructure import UnsignedInteger, UnsignedFloat, Integer, Typed
+from CADETProcess.optimization.optimizationProblem import OptimizationProblem
+from CADETProcess.optimization import OptimizerBase
+
+__all__ = ['AxInterface']
```
Comment (schmoelder) [inline CADETProcess/optimization/axAdapater.py:29]: Please update with AxInterface implementations. Remove AxInterface.

File: tests/test_optimizer_ax.py [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+except ImportError:
+    ax_imported = False
+
+import numpy as np
+import sys
+import unittest
+
+from CADETProcess.optimization import OptimizationProblem
+from CADETProcess.comparison import Comparator
+from CADETProcess.optimization.axAdapater import AxInterface
```
Comment (schmoelder) [inline tests/test_optimizer_ax.py:13]: Use concrete implementations of AxInterface. Include different tests. Also check in behaviour tests!

Comment (flo-schu) [reply tests/test_optimizer_ax.py:13]: This script is from the beginning of my work with Ax. I think we could use parts of it for system integration tests, but maybe the whole file is not necessary any longer. Let's discuss this.

File: tests/test_optimizer_ax.py [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+from matplotlib import pyplot as plt
+from scipy.signal import peak_widths, find_peaks
+
+
+sys.path.insert(0, '../')
+
+
+def create_parameters():
```
[the remark this answers is not in the document]
Comment (flo-schu) [reply tests/test_optimizer_ax.py:21]: yes

File: CADETProcess/optimization/__init__.py [hunk 4]
```diff
@@ -83,3 +90,33 @@
...
 from .scipyAdapter import COBYLA, TrustConstr, NelderMead, SLSQP
 from .pymooAdapter import NSGA2, U_NSGA3
+
+import importlib
+
+try:
+   from .axAdapater import GPEI, BotorchModular, NEHVI
+   from botorch.models.gp_regression import FixedNoiseGP
+   from botorch.acquisition import (
+      qExpectedImprovement,
```
Comment (schmoelder) [inline CADETProcess/optimization/__init__.py:100]: Why are these modules imported here? Should they be exposed to CADET-Process users?

File: CADETProcess/optimization/axAdapater.py [hunk 5]
```diff
@@ -0,0 +1,503 @@
...
+
+        F = obj_fun(X, untransform=True, n_cores=n_cores)
+
+        # Calculate nonlinear constraints
+        # TODO: add noise to the result
+        if self.optimization_problem.n_nonlinear_constraints > 0:
+            nonlincon_fun = self.optimization_problem.evaluate_nonlinear_constraints_population
+            nonlincon_labels = self.optimization_problem.nonlinear_constraint_labels
+
+            G = nonlincon_fun(X, untransform=True, n_cores=n_cores)
```
Comment (schmoelder) [inline CADETProcess/optimization/axAdapater.py:112]: update to use parallelization backend

e.g. 
```
nonlincon_fun(X, untransform=True, parallelization_backend=self.parallelization_backend)
```

**Comment by flo-schu** [comment]
tests are mostly running.

- x0 needs to be implemented for multi-objective tests. This could also improve the convergence speed
- test_if_solved method needs to be implemented for `Rosenbrock` and `LinearConstraintsSooTestProblem2`
- Only Moo tests are currently Failing, because the `test_if_solved` methods are a bit strict. Perhaps it would be enough to show that most points lie on the pareto front and not all. Otherwise, the optimizer needs to run for a long time, which is impractical.
- In addition, NEHVI + single obejctive raises an error that is catched. I added an inline todo which asks for a `supports_single_objective` flag.

Current test output
```
================================================================================================== short test summary info ==================================================================================================
FAILED tests/test_optimizer_behavior.py::test_convergence[LinearConstraintsMooTestProblem-optimizer2] - AssertionError: 
FAILED tests/test_optimizer_behavior.py::test_convergence[NonlinearConstraintsMooTestProblem-optimizer2] - AssertionError: 
FAILED tests/test_optimizer_behavior.py::test_from_initial_values[LinearConstraintsMooTestProblem-optimizer2] - NotImplementedError
FAILED tests/test_optimizer_behavior.py::test_from_initial_values[NonlinearConstraintsMooTestProblem-optimizer2] - NotImplementedError
FAILED tests/test_optimizer_behavior.py::test_resume_from_checkpoint[LinearConstraintsMooTestProblem-optimizer2] - assert 1 == 3
FAILED tests/test_optimizer_behavior.py::test_resume_from_checkpoint[NonlinearConstraintsMooTestProblem-optimizer2] - assert 2 == 3
============================================================================= 6 failed, 16 passed, 23 skipped, 39 warnings in 475.83s (0:07:55) =============================================================================
```

**Comment by flo-schu** [comment]
## Summary

**Ax runs quite well**. After an update GPEI uses LogExpectedImprovement which is much better at exploring small improvements. NEHVI can solve multiobjective problems, although the tests take some time to converge. But I think this feature will soon be mature enough for merging.

## Details

**Commits are granular. They should be very informative.**

**w.r.t: quick-n-dirty fix `Ax import`** @schmoelder you said you might have time to look into how this can be managed better. Feel free to do it on the branch. But this can also be done at the very end. But please no force-push :smile:

**batch mode vs. early stopping**: When I was doing the benchmark, I sacrificed batch mode in favor of a global early stopping strategy (they were incompatible, earlier commits). I wonder, if we can use early stopping strategies of the CADET post-processing, because the ax early stopping seems not to have an effect on multiobjective optimization. @schmoelder what do you think?

**Looking at Trieste**: For the time being it is clear that we are going to stick with Ax, but a breaking API in `BotorchModular` in the migration from `0.3.2 --> 0.3.5` made me looking at the documentation of `Trieste` and this is looking really great (and stable) + the API looks much less obscure than Ax's. 

**Next steps :hammer:** 
- Complete Ax tests, 
- Test other optimizers, 
- investigate bounds error of `COBYLA`, 
- increase speed of testing. @schmoelder Testing takes a long time, because of the many combinations between optimizers and tests. Also some optimizers are too slow for the simple tests. And increasing testing speed would also help in developing solutions. Perhaps you can also think about this. We could a) use a lower precision of the tests, b) implement some test like `approaches_true_solution` c) just live with it and tweak optimizers for the respective tests so they converge fast (e.g.) provide good initial values.

**Comment by schmoelder** [comment]
Hey @flo-schu,

thanks for all of your efforts. :pray: 

> w.r.t: quick-n-dirty fix Ax import @schmoelder you said you might have time to look into how this can be managed better. Feel free to do it on the branch. But this can also be done at the very end. But please no force-push 😄

I updated the relevant section. Hope it now works as intended. No force push, I promise! :nerd_face: 

> batch mode vs. early stopping: When I was doing the benchmark, I sacrificed batch mode in favor of a global early stopping strategy (they were incompatible, earlier commits). I wonder, if we can use early stopping strategies of the CADET post-processing, because the ax early stopping seems not to have an effect on multiobjective optimization. @schmoelder what do you think?

Tbh, I don't fully recall the difference here. Currently, there are no early stopping mechanisms in CADET-Process. However, they might be implemented in the future.

> Looking at Trieste: For the time being it is clear that we are going to stick with Ax, but a breaking API in BotorchModular in the migration from 0.3.2 --> 0.3.5 made me looking at the documentation of Trieste and this is looking really great (and stable) + the API looks much less obscure than Ax's.

Yeah, Ax seemed quite nice at first glance, especially since it's build on `torch`, but the interface has always been so clunky. I'm open to also exploring trieste as an alternative. "It's just another optimizer".

**Comment by flo-schu** [comment]
Final Todos

- [x] Test TrustConstr
- [x] Test SLSQP
- [ ] Test U_NSGA3
  - [ ] LinearEqualityConstraintsSooTestProblem --> `Cannot find individuals that fulfill constraints`
  - [ ] Moo Tests --> Accuracy too low. Reduce Margin here
  - [x] All other tests pass
- [ ] Test GPEI
- [ ] Test NEHVI

**Comment by flo-schu** [comment]
Locally all tests are passing (11 min). 

- I have made a minor change to the OptimizerBase (added supports_single_objective flag) and added a corresponding check. Please review.

Other than that all should be okay for now and ready for merge :tada:

File: CADETProcess/optimization/axAdapater.py [hunk 6]
```diff
@@ -0,0 +1,504 @@
...
+    OptimizationConfig, MultiObjectiveOptimizationConfig,
+    SearchSpace, MultiObjective, Experiment, Arm,
+    ComparisonOp, OutcomeConstraint,
+    RangeParameter, ParameterType, ParameterConstraint, Objective,
+
+)
+from ax.core.metric import MetricFetchResult, MetricFetchE
+from ax.core.base_trial import BaseTrial
+from ax.models.torch.botorch_modular.surrogate import Surrogate
+from ax.modelbridge.registry import Models
```
[the remark this answers is not in the document]
Comment (schmoelder) [reply CADETProcess/optimization/axAdapater.py:20]: please check again.

File: CADETProcess/optimization/axAdapater.py [hunk 7]
```diff
@@ -0,0 +1,562 @@
...
+        # This is interesting for logging but not quite yet what I'm looking for.
+        # I'd like to get the full "pareto" (i.e. non-dominated) front at that point.
+        # Try: exp_to_df!
+        # self._data = ax.Data.from_multiple_data([self._data, trial.fetch_data()])
+        # new_value = trial.fetch_data().df["mean"].min()
+
+        # print(
+        #     f"Iteration: Best in iteration {new_value:.3f}, "
+        #     f"Best so far: {self._data.df['mean'].min():.3f}"
+        # )
```
Comment (schmoelder) [inline CADETProcess/optimization/axAdapater.py:336]: If we don't need this anymore, please remove.

**Comment by schmoelder** [comment]
Great news, @flo-schu! Thanks for all your efforts!

I will try to run some problems locally to test performance of some "real" problems. Really excited!

Other than some minor remarks in the code base (see review comments), please make sure to clean up the git history a bit. I can also help you with that if you want.
Other than that, LGTM!

**Comment by schmoelder** [comment]
This branch was merged manually in e5cbf172649b2c63e5f8bdf25efc20159c62bc0d. Congrats @flo-schu and many thanks for this amazing contribution! :partying_face:

**Pull request closed without merging**